### PR TITLE
feat(turbo): task access trace & cached configs

### PR DESCRIPTION
**Pull request by tknickman**
### Description

v0 of work required for zero config (experimental):
1. Support config caching
2. Support task tracing

Requires: https://github.com/vercel/next.js/pull/59553

#### TODO
- [ ] Testing
- [ ] Support caching outputs on first run instead of second (turbo.json uncoupling)

### CODE REVIEW

File: crates/turborepo-lib/src/run/cache.rs [hunk 1]
```diff
@@ -374,3 +382,88 @@ impl TaskCache {
...
+    }
+
+    pub async fn restore(
+        &self,
+    ) -> Result<Option<(CacheHitMetadata, Vec<AnchoredSystemPathBuf>)>, CacheError> {
+        self.cache.fetch(&self.repo_root, &self.hash).await
+    }
+
+    pub async fn save(&self) -> Result<(), CacheError> {
+        match self.exists() {
```
Comment (chris-olszewski) [inline crates/turborepo-lib/src/run/cache.rs:427]: There's a TOCTOU race condition where the file gets cleaned up between us checking for it's existence and it actually getting read into memory.

File: crates/turborepo-lib/src/run/task_access.rs [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+fn task_access_trace_enabled(repo_root: &AbsoluteSystemPathBuf) -> Result<bool, Error> {
+    // TODO: use the existing config methods here
+    let root_turbo_json_path = &repo_root.join_component(TURBO_CONFIG_FILE);
+    if root_turbo_json_path.exists() {
+        return Ok(false);
+    }
+
+    // read package.json at root
+    let package_json_path = repo_root.join_components(&["package.json"]);
```
Comment (chris-olszewski) [inline crates/turborepo-lib/src/run/task_access.rs:62]: Is this feature only intended to be used in single package workspaces?

File: crates/turborepo-lib/src/task_graph/visitor.rs [hunk 3]
```diff
@@ -838,21 +860,49 @@ impl ExecContext {
...
+                        {
+                            error!("error caching output: {e}");
+                        } else {
+                            // If no errors, update hash tracker with expanded outputs
+                            self.hash_tracker.insert_expanded_outputs(
+                                self.task_id.clone(),
+                                self.task_cache.expanded_outputs().to_vec(),
+                            );
+                        }
+                    }
```
Comment (chris-olszewski) [inline crates/turborepo-lib/src/task_graph/visitor.rs:902]: This "can we cache" logic feels like it should belong totally in the task access module:
```
    // Whether we can cache the given task, returning None if task access isn't
    // enabled or the trace can't be found
    pub fn can_cache(&self, task_hash: &str, task_id: &str) -> Option<bool> {
        if !self.is_enabled() {
            return None;
        }
        let trace = TaskAccessTraceFile::read(&self.repo_root, task_hash)?;
        if trace.can_cache(&self.repo_root) {
            self.save_trace(trace, task_id.to_string());
            Some(true)
        } else {
            Some(false)
        }
    }
```
Which simplifies the changes in Visitor to
```suggestion
                   if self
                        .task_access
                        .can_cache(&self.task_hash, &self.task_id_for_display)
                        .unwrap_or(true)
                    {
                        if let Err(e) = self
                            .task_cache
                            .save_outputs(&mut prefixed_ui, task_duration, telemetry)
                            .await
                        {
                            error!("error caching output: {e}");
                        } else {
                            // If no errors, update hash tracker with expanded outputs
                            self.hash_tracker.insert_expanded_outputs(
                                self.task_id.clone(),
                                self.task_cache.expanded_outputs().to_vec(),
                            );
                        }
                    }
```

File: crates/turborepo-lib/src/task_graph/visitor.rs [hunk 4]
```diff
@@ -838,21 +860,49 @@ impl ExecContext {
...
+                    if self.task_access.is_enabled() {
+                        // Task access enabled: Handle trace file processing
+                        if let Some(trace) =
+                            TaskAccessTraceFile::read(&self.task_access.repo_root, &self.task_hash)
+                        {
+                            // Trace file found
+                            if trace.can_cache(&self.task_access.repo_root) {
+                                // Cache trace and save task outputs if trace is cacheable
+                                self.task_access
+                                    .save_trace(trace, self.task_id_for_display.to_string());
```
Comment (chris-olszewski) [inline crates/turborepo-lib/src/task_graph/visitor.rs:878]: Just making sure my understanding is correct: we want this to ensure we're getting a valid `TaskName` that can appear in a pipeline definition so we're using `task_id_for_display` as it'll end up being a valid pipeline task name?

**Review by chris-olszewski — CHANGES_REQUESTED** [review]
Mostly concerned about how this works with multi-package workspaces.

File: crates/turborepo-lib/src/run/cache.rs [hunk 5]
```diff
@@ -32,11 +36,15 @@ pub enum Error {
...
+    #[error(transparent)]
+    Scm(#[from] turborepo_scm::Error),
+    #[error(transparent)]
+    Path(#[from] turbopath::PathError),
 }
 
 pub struct RunCache {
     task_output_mode: Option<OutputLogsMode>,
-    cache: AsyncCache,
+    cache: Arc<AsyncCache>,
```
[the remark this answers is not in the document]
Comment (tknickman) [reply crates/turborepo-lib/src/run/cache.rs:47]: Good call, broke out that PR into it's own: https://github.com/vercel/turbo/pull/7293

File: crates/turborepo-lib/src/run/task_access.rs (hunk 2 above)
Comment (tknickman) [reply crates/turborepo-lib/src/run/task_access.rs:62]: For now, yes (but we'll roll this out more broadly soon)

File: crates/turborepo-lib/src/task_graph/visitor.rs (hunk 4 above)
Comment (tknickman) [reply crates/turborepo-lib/src/task_graph/visitor.rs:891]: Yep that's correct, a little weird in this case, but we need that to build the pipeline later on.

**Comment by tknickman** [comment]
@chris-olszewski PR updated.

There were a lot of conflicts here, so split the commits between resolving conflicts, and addressing review. But let me know if you have any questions!

NOTE: This is stacked on top of https://github.com/vercel/turbo/pull/7293

File: crates/turborepo-lib/src/run/cache.rs (hunk 1 above)
Comment (chris-olszewski) [reply crates/turborepo-lib/src/run/cache.rs:418]: Tracing through this, I don't think there's an easy way around it unless we rework the cache abstraction a fair bit. I think it's fine to leave as is 🤷

**Review by chris-olszewski — APPROVED** [review]
LFG 🚀

**Pull request merged**